extras/FvwmConfig/FvwmConfig.man
.\" t
.\" @(#)FvwmConfig.1	1/12/94
.TH FvwmConfig 1.20 "Mar 16 1994"
.UC
.SH NAME
FvwmConfig \- the FVWM Configuration Aid (Braindead)
.SH SYNOPSIS
FvwmConfig is spawned by fvwm, so no command line invocation will work.
From within the .fvwmrc file, FvwmConfig is spawned as follows:
.nf
.sp
Module FvwmConfig
.sp
.fi
or from within an fvwm pop-up menu:
.nf
.sp
     AddToMenu Module-Popup "Modules" Title
     +         "Configure"      Module FvwmConfig
     +         "FvwmIdentify"   Module FvwmIdent
     +         "Banner"         Module FvwmBanner
     +         "SaveDesktop"    Module FvwmSave
.sp
.fi

.SH DESCRIPTION
The FvwmConfig helps users make very simple, basic changes to the
current fvwm configuration.

In order for FvwmConfig to make changes which will take effect for
future X sessions, the user's or system's .fvwmrc file must
end with the line "Read .fvwmrc-config". FvwmConfig will write its
configuration out to the file $HOME/.fvwmrc-config, thus minimizing
conflicts with other portions of the user's configuration.

Since FvwmConfig is used to edit the user's configuration, it doesn't
make much sense for it to be very configurable itself. There are no
user-specifiable configuration options for FvwmConfig.

Use of this module is discouraged, since the options it allows are
far less in variety than fvwm itself can allow.

FvwmConfig will not work as intended if ANY styles other than the
default styles are applied to it.

.SH COPYRIGHTS
The FvwmConfig program, and the concept for interfacing this module to
the Window Manager, are all original work by Robert Nation.

Copyright 1994, Robert Nation. No guarantees or warranties or anything
are provided or implied in any way whatsoever. Use this program at your
own risk. Permission to use this program for any purpose is given,
as long as the copyright is kept intact. 


.SH AUTHOR
Robert Nation 



NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in stable release 2.6.8 (UNRELEASED)

* Bug fixes:

  - Various DESTDIR fixes (especially around the default-config
    Makefile)
  - fvwm-perllib pod2man fixes.
  - FvwmIconMan no longer triggers a warning about bad size hints
    in the fvwm core.
  - VMware windows do not disapper when switching between
    fullscreen and normal state.
  - Fix (de)installation with the configure options
    --program-prefix, --program-suffix and
    --program-transform-name.

-------------------------------------------------------------------
Changes in stable release 2.6.7 (06-Mar-2016)

* New fvwm features:

  - A new default configuration which is available when fvwm
    doesn't detect a configuration file to load.
  - A new conitional command "Desk n" can restrict matching
    windows to a specific desk.
  - A new conditional command "Screen n" to restrict matching
    windows on a given Xinerama screen.
  - New expansion variable "w.screen" to ascertain the Xineram
    screen number a window is on.
  - New command "InfoStoreClear" to remove all items in the
    InfoStore.

* Removed features:

  - The old and unmaintained debian/and rpm/ directories have
    been remmoved; use the maintainers' copies where available.
  - VMS support has been removed.
  - GTK1.x support has been removed.
  - GNOME-specific window hints (pre-EWMH) have been removed.
  - Some fvwm modules have been removed:
      - FvwmDragWell   (no replacement)
      - FvwmGTK        (no replacement)
      - FvwmSave       (no replacement)
      - FvwmSaveDesk   (no replacement)
      - FvwmScroll     (no replacement)
      - FvwmTabs       (no replacement, never worked anyway)
      - FvwmTaskBar    (use FvwmButtons)
      - FvwmTheme      (in core of fvwm as colorsets)
      - FvwmWharf      (use FvwmButtons)
      - FvwmWinList    (use WindowList command)
      - FvwmWindowMenu (use WindowList command)
      - FvwmIconBox    (use the IconBox style instead)

* New module features:

  - FvwmButtons learned a new option "Colorset" to its
    ChangeButton command.

* fvwm-menu-desktop updated:

  - Renamed default menu to XDGMenu and changed the name
    of the FvwmForm to FvwmForm-XDGMenu-Config to not conflict
    with someone already using FvwmMenu.
  - fvwm-menu-desktop will now load defaults from the
    FvwmForm-XDGMenu-Config data file.
  - Improved dynamic menus by regenerating them on-the-fly.
  - Added new options: --regen-cmd, --dynamic, and more.

* Bug fixes:

  - A bug introduced in 2.6.6 could cause applications with
    negative coordinates to be placed at strange positions.  This
    affected for example acroread when switching to fullscreen
    mode.  This has been fixed.
  - FvwmButtons "Silent" option for dynamic ChangeButton commands
    no longer loops infinitely.

Changes in stable release 2.6.6 (15-Mar-2016)

* New fvwm features:

  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.
  - The new extended variables $[wa.x], $[wa.y], $[wa.width],
    $[wa.height] can be used to get the geometry of the EWMH
    working area, and $[dwa.x], $[dwa.y], $[dwa.width],
    $[dwa.height] can be used to get the geometry of the EWMH
    dynamic working area.
  - The Resize commands accept "wa" or "da" as a suffix of the
    width or height value.  If present, the value is a percentage
    of the width or height of the EWMH working area or the EWMH
    dynamic working area.
  - Fvwm is much more resilient against applications that flood the
    window manager with repeated events.

* New module features:

  - FvwmForm supports separator lines.
  - New FvwmIconMan options:
      IconAndSelectButton
      IconAndSelectColorset

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-menu-desktop is re-written and provides better support of
    the XDG menu specification.
  - Fix fvwm-menu-desktop keyError bug.  Use "others" if no desktop
    environment found.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take focus themselves (style
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.
  - FvwmConsole now causes much less network traffic.
  - Suppress bogus events sent to the modules when a window is
    resized with the mouse.
  - Properly handle the has_ref_window_moved flag for ResizeMove and
    ResizeMoveMaximize.
  - Removed some unnecessary redraws in FvwmPager.
  - The option "NoDeskLimitY" option of the GotoPage command did
    not work.
  - Negative coordinates in the "rectangle" option to the Menu
    commend did not work correctly.  This has been fixed.
  - Removes a slight graphics problem whith the ResizeMaximize
    command being invoked from a window button menu.
  - When an attempt to reparent a client window (i.e. decorate
    it) fails, fvwm no longer throws away all events but only the
    events for that window.

extras/FvwmConfig/FvwmConfig.C
#include "config.h"

#include <stdio.h>
#include <iostream.h>
#include <math.h>

#include "WinBase.h"
#include "WinButton.h"
#include "WinButton.h"
#include "WinInput.h"
#include "WinColorSelector.h"

extern "C" {
#include <libs/fvwmlib.h>
}
#include <fvwm/module.h>

#define ClickToFocus (1<<10)
#define SloppyFocus (1<<11)
int fd[2];
char *MyName;

extern "C" void SetupPipeHandler(void);

 
WinInput *inwin;
WinButton * cfocus, *cmouse,*fmouse, *fclick, *fsloppy, *nescroll, *escroll;
WinColorSelector *hibackselwin, *hiforeselwin, *backselwin, *foreselwin;
WinBase *base;
Window pwin;

void finish(WinBase *which)
{
  exit(0);
}

void abortaction(int newstate, WinButton *which)
{
  exit(0);
}

char tmp[2000];
int focus = -1;
int colormap = -1;
int edgescroll = -1;
float back_red = -1;
float back_green = -1;
float back_blue = -1;
float fore_red = -1;
float fore_green = -1;
float fore_blue = -1;

float hi_back_red = -1;
float hi_back_green = -1;
float hi_back_blue = -1;
float hi_fore_red = -1;
float hi_fore_green = -1;
float hi_fore_blue = -1;
int size_read = -1;


void testaction(int newstate, WinButton *which)
{
  int i;
  char *t;

  if(focus == 1)
    SendInfo(fd,"Style \"*\" ClickToFocus\n",0);
  else if(focus == 0)
    SendInfo(fd,"Style \"*\" MouseFocus\n",0);
  else if(focus == 2)
    SendInfo(fd,"Style \"*\" SloppyFocus\n",0);    

  if(colormap == 1)
    SendInfo(fd,"ColormapFocus FollowsFocus",0);
  else if (colormap == 0)
    SendInfo(fd,"ColormapFocus FollowsMouse",0);

  t = inwin->GetLine();
  if(strlen(inwin->GetLine()) > 2)
    {
      if(sscanf(t,"%dx%d", &i, &i) == 2)
	{
	  sprintf(tmp,"DeskTopSize %s\n",t);
	  SendInfo(fd,tmp,0);
	}
      else
	{
	  inwin->SetLabel("");
	  inwin->RedrawWindow(1);
	}
    }
  else
    {
      inwin->SetLabel("");
      inwin->RedrawWindow(1);
    }

  if(fore_red >= 0)
    {
      sprintf(tmp,"Style \"*\" ForeColor #%04x%04x%04x",(int)fore_red,
	      (int)fore_green, (int)fore_blue);
      SendInfo(fd,tmp,0);
    }
  if(back_red >= 0)
    {
      sprintf(tmp,"Style \"*\" BackColor #%04x%04x%04x",(int)back_red,
	      (int)back_green, (int)back_blue);
      SendInfo(fd,tmp,0);
    }

  if((hi_fore_red >= 0)||(hi_back_red >= 0))
    {
      sprintf(tmp,"HilightColor #%04x%04x%04x #%04x%04x%04x",(int)hi_fore_red,
	      (int)hi_fore_green, (int)hi_fore_blue,(int)hi_back_red,
	      (int)hi_back_green, (int)hi_back_blue);
      SendInfo(fd,tmp,0);
    }
  if(edgescroll == 1)
    SendInfo(fd,"EdgeScroll 100 100",0);
  else if (edgescroll == 0)
    SendInfo(fd,"EdgeScroll 0 0",0);	     
  
  if((focus == 1)||(focus == 0)||(fore_red >= 0)||(back_red >=0))
    SendInfo(fd,"Recapture",0);
}

void commitaction(int newstate, WinButton *which)
{
  char fname[2000];
  char *homedir;
  FILE *nfd;
  char *t;
  int i;

  testaction(newstate, which);
  homedir = getenv("HOME");
  if(homedir == NULL)
    fname[0] = 0;
  else
    strcpy(fname,homedir);
  strcat(fname,"/.fvwmrc-config");

  nfd = fopen(fname,"w");
  if(nfd)
    {
      if(focus == 1)
	fprintf(nfd,"Style \"*\" ClickToFocus\n");
      else if(focus == 0)
	fprintf(nfd,"Style \"*\" MouseFocus\n");
      else if(focus == 2)
	fprintf(nfd,"Style \"*\" SloppyFocus\n");          

      if(colormap == 1)
	fprintf(nfd,"ColormapFocus FollowsFocus\n");
      else if (colormap == 0)
	fprintf(nfd,"ColormapFocus FollowsMouse\n");
      
      t = inwin->GetLine();
      if(strlen(inwin->GetLine()) > 2)
	{
	  if(sscanf(t,"%dx%d", &i, &i) == 2)
	    fprintf(nfd,"DeskTopSize %s\n",inwin->GetLine());
	}
      
      if(edgescroll == 1)
	fprintf(nfd,"EdgeScroll 100 100\n");
      else if (edgescroll == 0)
	fprintf(nfd,"EdgeScroll 0 0\n");	     
      if(fore_red >= 0)
	{
	  fprintf(nfd,"Style \"*\" ForeColor #%04x%04x%04x\n",(int)fore_red,
		  (int)fore_green, (int)fore_blue);
	}
      if(back_red >= 0)
	{
	  fprintf(nfd,"Style \"*\" BackColor #%04x%04x%04x\n",(int)back_red,
		  (int)back_green, (int)back_blue);
	}      
      if((hi_fore_red >= 0)||(hi_back_red >= 0))
	{
	  fprintf(nfd,"HilightColor #%04x%04x%04x #%04x%04x%04x\n",
		  (int)hi_fore_red,(int)hi_fore_green, (int)hi_fore_blue,
		  (int)hi_back_red, (int)hi_back_green, (int)hi_back_blue);
	}
      fclose(nfd);
    }
  exit(0);
}

void backcolorhandler(float red, float green, float blue, 
		      WinColorSelector *which)
{
  back_red = red;
  back_green = green;
  back_blue = blue;
}

void forecolorhandler(float red, float green, float blue, 
		     WinColorSelector *which)
{
  fore_red = red;
  fore_green = green;
  fore_blue = blue;
}

void hibackcolorhandler(float red, float green, float blue, 
		      WinColorSelector *which)
{
  hi_back_red = red;
  hi_back_green = green;
  hi_back_blue = blue;
}

void hiforecolorhandler(float red, float green, float blue, 
		     WinColorSelector *which)
{
  hi_fore_red = red;
  hi_fore_green = green;
  hi_fore_blue = blue;
}

void  colormapmouseaction(int newstate, WinButton *which)
{
  if(newstate == 1)
    {
      if(colormap == 0)
	colormap = -1;
    }
  else
    {
      cfocus->PopOut(); 
      colormap = 0;
      cfocus->RedrawWindow(0);
    }
 
}

void  Scrollaction(int newstate, WinButton *which)
{
  if(newstate == 1)
    {
      if(edgescroll == 1)
	edgescroll = -1;
    }
  else
    {
      nescroll->PopOut();
      edgescroll = 1;
      nescroll->RedrawWindow(0);
    }
}
void  Noscrollaction(int newstate, WinButton *which)
{
  if(newstate == 1)
    {
      if(edgescroll == 0)
	 edgescroll= -1;
    }
  else
    {
      escroll->PopOut();
      edgescroll = 0;
      escroll->RedrawWindow(0);
    }
}
void  colormapfocusaction(int newstate, WinButton *which)
{
  if(newstate == 1)
    {
      if(colormap==1)
	colormap = -1;
    }
  else
    {
      cmouse->PopOut(); 
      colormap = 1;
      cmouse->RedrawWindow(0);
    }
}

void  focusmouseaction(int newstate, WinButton *which)
{
  if(newstate == 1)
    {
      if(focus == 0)
      focus = -1;
    }
  else
    {
      fclick->PopOut();
      focus = 0;
      fclick->RedrawWindow(0);
      fsloppy->PopOut();
      fsloppy->RedrawWindow(0);
    }
}
void  focusclickaction(int newstate, WinButton *which)
{
  if(newstate == 1)
    {
      if(focus == 1)
	focus = -1;
    }
  else
    {
      fmouse->PopOut();
      focus = 1;
      fmouse->RedrawWindow(0);
      fsloppy->PopOut();
      fsloppy->RedrawWindow(0);
    }
}
void  focussloppyaction(int newstate, WinButton *which)
{
  if(newstate == 1)
    {
      if(focus == 2)
	focus = -1;
    }
  else
    {
      focus = 2;
      fmouse->PopOut();
      fmouse->RedrawWindow(0);
      fclick->PopOut();
      fclick->RedrawWindow(0);
    }
}

void ReadPacket(int fd)
{
  unsigned long header[HEADER_SIZE];
  unsigned long *body;
  Pixel hipixel, hibackpixel;
  XColor colorcell;
  int n;

  ReadFvwmPacket(fd, header, &body);
  
  if(header[1] == M_FOCUS_CHANGE)
    {
      if(hi_fore_red < 0)
	{
	  hipixel = body[3];
	  hibackpixel = body[4];
	  colorcell.pixel = hipixel;
	  n = XQueryColor(hiforeselwin->dpy, 
			  DefaultColormap(hiforeselwin->dpy,
					  hiforeselwin->Screen),
			  &colorcell);
	  hi_fore_red = colorcell.red & 0xffff;
	  hi_fore_green = colorcell.green & 0xffff;
	  hi_fore_blue = colorcell.blue & 0xffff;
	  hiforeselwin->SetCurrentValue(hi_fore_red, hi_fore_green, hi_fore_blue);
	  colorcell.pixel = hibackpixel;
	  XQueryColor(hibackselwin->dpy, 
		      DefaultColormap(hiforeselwin->dpy,
				      hiforeselwin->Screen),
		      &colorcell);
	  hi_back_red = colorcell.red & 0xffff;
	  hi_back_green = colorcell.green & 0xffff;
	  hi_back_blue = colorcell.blue & 0xffff;
	  hibackselwin->SetCurrentValue(hi_back_red, hi_back_green, hi_back_blue);
	}
    }
  else if((header[1] == M_CONFIGURE_WINDOW)&&(body[0] == pwin))
    {
      if(fore_red < 0)
	{
	  hipixel = body[22];
	  hibackpixel = body[23];
	  colorcell.pixel = hipixel;
	  XQueryColor(foreselwin->dpy,
		      DefaultColormap(hiforeselwin->dpy,
				      hiforeselwin->Screen),
		      &colorcell);
	  fore_red = colorcell.red & 0xffff;
	  fore_green = colorcell.green & 0xffff;
	  fore_blue = colorcell.blue & 0xffff;
	  foreselwin->SetCurrentValue(fore_red, fore_green, fore_blue);
	  colorcell.pixel = hibackpixel;
	  XQueryColor(backselwin->dpy,
		      DefaultColormap(hiforeselwin->dpy,
				      hiforeselwin->Screen),
		      &colorcell);
	  back_red = colorcell.red & 0xffff;
	  back_green = colorcell.green & 0xffff;
	  back_blue = colorcell.blue & 0xffff;
	  backselwin->SetCurrentValue(back_red, back_green, back_blue);      
	  if(body[8] & ClickToFocus)
	    focus = 1;
	  if(body[8] & SloppyFocus)
	    focus = 2;
	  if(focus == 1)
	    {
	      fclick->PushIn();
	      fclick->RedrawWindow(0);
	      fmouse->PopOut();
	      fmouse->RedrawWindow(0);
	      fsloppy->PopOut();
	      fsloppy->RedrawWindow(0);
	    }
	  else if (focus == 2)
	    {
	      fclick->PopOut();
	      fclick->RedrawWindow(0);
	      fmouse->PopOut();
	      fmouse->RedrawWindow(0);
	      fsloppy->PushIn();
	      fsloppy->RedrawWindow(0);
	    }
	  else
	    {
	      fclick->PopOut();
	      fclick->RedrawWindow(0);
	      fmouse->PushIn();
	      fmouse->RedrawWindow(0);
	      fsloppy->PopOut();
	      fsloppy->RedrawWindow(0);
	    }
	  
	}
    }
  else if(header[1] == M_NEW_PAGE)
    {
      char size[100];
      int vxmax, vymax, sx,sy,nx,ny;

      if(size_read < 0)
	{
	  vxmax = body[3];
	  vymax = body[4];
	  sx = base->ScreenWidth();
	  sy = base->ScreenHeight();
	  nx = vxmax/sx + 1;
	  ny = vymax/sy + 1;
	  sprintf(size,"%dx%d",nx,ny);
	  inwin->SetLabel(size);    
	  size_read = 1;
	}
    }
  free((char *)body);
}

#define WINDOW_HEIGHT 400
int main(int argc, char **argv)
{
  char *temp, *s;

  /* Save our program  name - for error messages */
  temp = argv[0];
  s=strrchr(argv[0], '/');
  if (s != NULL)
    temp = s + 1;
  
  MyName = safemalloc(strlen(temp)+2);
  strcpy(MyName, temp);
  
  if(argc  < 6)
    {
      fprintf(stderr,"%s Version %s should only be executed by fvwm!\n",MyName,
	      VERSION);
      exit(1);
    }
  
  /* Dead pipe == Fvwm died */
  SetupPipeHandler();
  fd[0] = atoi(argv[1]);
  fd[1] = atoi(argv[2]);

  WinInitialize(argv,argc);
  WinAddInput(fd[1],ReadPacket);
  SendInfo(fd,"Send_WindowList",0);
  WinBase a(NULL,300,WINDOW_HEIGHT,0,0);
  pwin = a.win;
  base = &a;
  a.SetWindowName("Braindead Fvwm Configuration Editor");
  a.SetIconName("Braindead");
  a.SetBevelWidth(3);
  a.SetGeometry((a.ScreenWidth() - 200)/2,(a.ScreenHeight()-200)/2,
		300,WINDOW_HEIGHT,
		1,1,500,500,1,1,1,1,CenterGravity); 

  WinText keyboard(&a,280,20,10,6,"Keyboard Focus");
  keyboard.SetBevelWidth(0);
  WinButton focusclick(&a,90,20,15,32,"Click");
  WinButton focusmouse(&a,90,20,105,32,"Follows Mouse");
  WinButton focussloppy(&a,90,20,195,32,"Trails Mouse");
  focusmouse.SetToggleAction(focusmouseaction);
  focusclick.SetToggleAction(focusclickaction);
  focussloppy.SetToggleAction(focussloppyaction);

  WinBase divider1(&a,290,4,5,58);
  divider1.SetBevelWidth(2);
  WinText colormap(&a,280,20,10,62,"Colormap Focus");
  colormap.SetBevelWidth(0);
  WinButton colormapfocus(&a,100,20,50,88,"Follows Focus");
  WinButton colormapmouse(&a,100,20,150,88,"Follows Mouse");
  colormapmouse.SetToggleAction(colormapmouseaction);
  colormapfocus.SetToggleAction(colormapfocusaction);

  WinBase divider2(&a,290,4,5,114);
  divider2.SetBevelWidth(2);
  WinText desktop(&a,280,20,10,120,"Desktop");
  desktop.SetBevelWidth(0);
  WinText desktopsize(&a,30,20,20,144,"Size");
  desktopsize.SetBevelWidth(0);
  WinInput deskinput(&a,40,20,50,144,NULL);
  deskinput.SetBevelWidth(2);

  WinButton EdgeScroll(&a,80,20,110,144,"Edge Scroll");
  WinButton NoEdgeScroll(&a,100,20,190,144,"No Edge Scroll");
  EdgeScroll.SetToggleAction(Scrollaction);
  NoEdgeScroll.SetToggleAction(Noscrollaction);

  WinBase divider3(&a,290,4,5,170);
  divider3.SetBevelWidth(2);
  WinText backcolorlabel(&a,140,20,5,176,"Default Back Color");
  WinText forecolorlabel(&a,140,20,155,176,"Default Fore Color");
  WinColorSelector defaultbackcolor(&a,140,70,5,196);
  WinColorSelector defaultforecolor(&a,140,70,155,196);
  backselwin = &defaultbackcolor;
  foreselwin = &defaultforecolor;
  defaultbackcolor.SetBevelWidth(0);
  defaultforecolor.SetBevelWidth(0);
  backcolorlabel.SetBevelWidth(0);
  forecolorlabel.SetBevelWidth(0);
  defaultbackcolor.SetMotionAction(backcolorhandler);
  defaultforecolor.SetMotionAction(forecolorhandler);

  WinBase divider4(&a,290,4,5,272);
  divider4.SetBevelWidth(2);
  WinText hibackcolorlabel(&a,140,20,5,278,"Hilight Back Color");
  WinText hiforecolorlabel(&a,140,20,155,278,"Hilight Fore Color");

  WinColorSelector hibackcolor(&a,140,70,5,298);
  WinColorSelector hiforecolor(&a,140,70,155,298);
  hibackselwin = &hibackcolor;
  hiforeselwin = &hiforecolor;
  hibackcolor.SetBevelWidth(0);
  hiforecolor.SetBevelWidth(0);
  hibackcolorlabel.SetBevelWidth(0);
  hiforecolorlabel.SetBevelWidth(0);
  hibackcolor.SetMotionAction(hibackcolorhandler);
  hiforecolor.SetMotionAction(hiforecolorhandler);

  
  WinButton testbutton(&a,60,20,30,WINDOW_HEIGHT-30,"Test");
  WinButton commitbutton(&a,60,20,120,WINDOW_HEIGHT-30,"Commit");
  testbutton.MakeMomentary();
  WinButton abortbutton(&a,60,20,210,WINDOW_HEIGHT-30,"Abort");

  commitbutton.SetToggleAction(commitaction);
  testbutton.SetToggleAction(testaction);
  abortbutton.SetToggleAction(abortaction);
  fclick = &focusclick;
  fsloppy = &focussloppy;
  fmouse = &focusmouse;
  cmouse = &colormapmouse;
  cfocus = &colormapfocus;
  escroll = &EdgeScroll;
  nescroll = &NoEdgeScroll;
  inwin = &deskinput;
  a.Map();

  while(1)
    {
      WinLoop();
    }
  return 0;
}
